### PR TITLE
feat(testing): add port resolution infrastructure for environment-resilient tests

### PR DESCRIPTION
**Pull request by MacAttak**
## Summary

- Add `SERVICE_DEFAULT_PORTS` dict and `get_effective_port()` for environment-variable-based port resolution, mirroring the existing host resolution pattern
- Extend `ServiceEndpoint` with `__post_init__` port resolution and `.url` property so `ServiceEndpoint("polaris")` auto-resolves port 8181
- Update `check_infrastructure()` and `IntegrationTestBase.required_services` to accept both string and tuple formats (`["polaris"]` or `[("polaris", 8181)]`)
- Export canonical `{SERVICE}_PORT` env vars from `test-e2e.sh` after port-forward setup

## Acceptance Criteria

| AC | Description | Status |
|----|-------------|--------|
| AC-1 | `SERVICE_DEFAULT_PORTS` single source of truth (13 services) | PASS |
| AC-2 | `_get_effective_port()` precedence: env var → default → dict → ValueError | PASS |
| AC-3 | `get_effective_port()` publicly exported | PASS |
| AC-4 | `ServiceEndpoint.__post_init__` resolves port when omitted | PASS |
| AC-5 | `ServiceEndpoint.url` returns `http://{host}:{port}` | PASS |
| AC-6 | `check_infrastructure()` accepts `list[tuple[str, int] | str]` | PASS |
| AC-7 | `IntegrationTestBase` accepts both formats, optional port | PASS |
| AC-8 | `test-e2e.sh` exports DAGSTER_WEBSERVER_PORT, DAGSTER_PORT, MARQUEZ_PORT | PASS |
| AC-9 | Invalid env var produces actionable error with env var name | PASS |
| AC-10 | Backward compatibility — existing tests unchanged | PASS |

## Blast Radius

- **testing/fixtures/services.py**: Port resolution added alongside existing host resolution. Failure scope: local (only affects tests using new string format).
- **testing/base_classes/integration_test_base.py**: Type hint widened (`list` → `Sequence`), port made optional. Failure scope: adjacent (affects subclasses, but backward-compatible).
- **testing/ci/test-e2e.sh**: 3 new exports. Failure scope: local (new vars, no existing behavior changed).
- **NOT changed**: No production code, no existing test files, no Helm charts, no CI pipelines.

## Gate Results

| Gate | Status | Findings (B/W/I) |
|------|--------|-------------------|
| Build | PASS | 0/0/0 |
| Tests | PASS | 0/0/0 |
| Security | PASS | 0/0/0 |
| Wiring | PASS | 0/1/5 |
| Spec | PASS | 0/0/4 |

**Wiring WARN**: `get_effective_port` and `SERVICE_DEFAULT_PORTS` not yet re-exported from `testing/fixtures/__init__.py` (will be added in Unit 2: call-site-migration).

## Evidence

- 115 unit tests across 8 test classes (all passing)
- 6 commits, 4 files changed (+1894/-27 lines)
- Gate evidence: `.specwright/work/env-resilient-test-infra/units/port-resolution-infra/evidence/gate-results.md`

## Test Plan

- [x] `make test-unit` passes (115 new + all existing)
- [x] mypy --strict passes (Sequence covariance fix)
- [x] Pre-push hooks pass (ruff, bandit, mypy, pytest, helm lint, traceability)
- [ ] `make test-e2e` validates shell exports work end-to-end (requires Kind cluster)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: testing/fixtures/services.py [hunk 1]
```diff
@@ -26,8 +26,101 @@
...
+    Returns:
+        Resolved port as an integer.
+
+    Raises:
+        ValueError: If the env var contains a non-integer value, or if no
+            port can be determined.
+    """
+    env_key = f"{service_name.upper().replace('-', '_')}_PORT"
+    env_val = os.environ.get(env_key)
+    if env_val is not None and env_val != "":
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply testing/fixtures/services.py]: Fixed in e2b295b. Added `.strip()` before the emptiness check — whitespace-only env vars now fall through to defaults consistently.

File: testing/fixtures/services.py [hunk 2]
```diff
@@ -26,8 +26,101 @@
...
+        try:
+            return int(env_val)
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply testing/fixtures/services.py]: Fixed in e2b295b. Added TCP range validation (1-65535) after `int()` conversion with a clear error message. Also added 3 new tests: port 0, port >65535, and negative port.

File: testing/tests/unit/test_port_resolution.py [hunk 3]
```diff
@@ -0,0 +1,1743 @@
...
+        """
+        monkeypatch.setenv("POLARIS_PORT", "   ")
+        # Either falls through to dict default (8181) or raises ValueError
+        # It must NOT return 0 or some garbage value
+        try:
+            result = _get_effective_port("polaris")
+            # If it didn't raise, it must have fallen through to default
+            assert result == 8181
+        except ValueError:
+            pass  # Also acceptable - whitespace is not a valid integer
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply testing/tests/unit/test_port_resolution.py]: Fixed in e2b295b. With the `.strip()` change, whitespace-only values now deterministically fall through to defaults. The test asserts `result == 8181` with no `try/except` ambiguity.
